electrum/gui/qt/wizard/wallet.py

```diff
@@ -0,0 +1,1382 @@
+        client.handler = self.plugin.create_handler(self.wizard)
+
+        def unlock_task(client):
+            try:
+                self.password = client.get_password_for_storage_encryption()
+            except Exception as e:
+                self.error = repr(e)  # TODO: handle user interaction exceptions (e.g. invalid pin) more gracefully
+                self.logger.error(repr(e))
+            self.busy = False
+            self.validate()
+
+        t = threading.Thread(target=unlock_task, args=(client,), daemon=True)
+        t.start()
+
+    def validate(self):
+        if self.password and not self.error:
+            self.apply()
+            self.valid = True
+        else:
+            self.valid = False
+
+        if self.valid:
+            self.wizard.requestNext.emit()  # via signal, so it triggers Next/Finish on GUI thread after on_updated()
+
+    def apply(self):
+        if self.valid:
+            self.wizard_data['password'] = self.password
+
+
+class WCHWXPub(WizardComponent, Logger):
+    def __init__(self, parent, wizard):
+        WizardComponent.__init__(self, parent, wizard, title=_('Retrieving extended public key from hardware'))
+        Logger.__init__(self)
+        self.plugins = wizard.plugins
+        self.plugin = None
+        self._busy = True
+
+        self.xpub = None
+        self.root_fingerprint = None
+        self.label = None
+        self.soft_device_id = None
+
+        self.ok_l = WWLabel(_('Hardware keystore added to wallet'))
+        self.ok_l.setAlignment(Qt.AlignCenter)
+        self.layout().addWidget(self.ok_l)
+
+    def on_ready(self):
+        _name, _info = self.wizard_data['hardware_device']
+        self.plugin = self.plugins.get_plugin(_info.plugin_name)
+        self.title = _('Retrieving extended public key from {} ({})').format(_info.model_name, _info.label)
+
+        device_id = _info.device.id_
+        client = self.plugins.device_manager.client_by_id(device_id, scan_now=False)
+        if not client.handler:
+            client.handler = self.plugin.create_handler(self.wizard)
+
+        cosigner = self.wizard.current_cosigner(self.wizard_data)
+        xtype = cosigner['script_type']
+        derivation = cosigner['derivation_path']
+
+        def get_xpub_task(client, derivation, xtype):
+            try:
+                self.xpub = self.get_xpub_from_client(client, derivation, xtype)
+                self.root_fingerprint = client.request_root_fingerprint_from_device()
+                self.label = client.label()
+                self.soft_device_id = client.get_soft_device_id()
+            except UserFacingException as e:
+                self.error = str(e)
+                self.logger.error(repr(e))
+            except Exception as e:
+                self.error = repr(e)  # TODO: handle user interaction exceptions (e.g. invalid pin) more gracefully
+                self.logger.error(repr(e))
```

Comment:
ah was about to comment about this, but I see there is already this TODO.
With trezor1 and keepkey, on invalid PIN, the screen looks like this:
![pic3](https://github.com/spesmilo/electrum/assets/29142493/5564bb22-d0be-404d-a880-3b0efafe4301)

Reply:
These exceptions are coming from the device client libs. It would be unwieldy to catch them all within `WCHWXPub` class as it is a generic component used for all devices, so I think the better approach is that the (device specific) device client class catches these low level exceptions and re-raise them within a `UserFacingException`.